Namespace approval issues fix

helen229 opened the pull request:
close: https://github.com/Azure/azure-sdk-tools/issues/12211
close: https://github.com/Azure/azure-sdk-tools/issues/12204
close: https://github.com/Azure/azure-sdk-tools/issues/12215
close: https://github.com/Azure/azure-sdk-tools/issues/12644

Front-end: Added hasRequestedNamespaceReview flag to APIRevision model and integrated state handling in review page/options components.
UI logic: Adjusted enable/disable conditions and button state handling for namespace review requests when a review is approved or already requested.
Back-end: Added HasRequestedNamespaceReview persistence and expanded logging plus revision flag update during namespace review request flow.

tjprescott reviewed (changes requested):
> helen229 please add a description about what you are fixing, that would help us a lot with the review! also, if you can include more details on the title of the pull request, it would be great :)

Address all of these issues. These are minimum standards expected of all PRs.

helen229 commented on src/dotnet/APIView/APIViewWeb/LeanModels/ReviewListModels.cs:
```diff
@@ -148,6 +148,7 @@ public string PackageVersion
         public bool CopilotReviewInProgress { get; set; }
         public string CopilotReviewJobId { get; set; }
         public HashSet<string> Approvers { get; set; } = new HashSet<string>();
+        public bool HasRequestedNamespaceReview { get; set; } = false;
```
we use this property to check whether this revision has been requested review. This property is on revision level not in review model level, NamespaceReviewStatus is on review model level so we can't reuse it.

AlitzelMendez commented on src/dotnet/APIView/APIViewWeb/Managers/ReviewManager.cs:
```diff
@@ -406,23 +441,37 @@ private async Task MarkAssociatedReviewsForNamespaceReview(List<ReviewListItemMo
+                        else if (review != null && LanguageHelper.IsSDKLanguageOrTypeSpec(review.Language) && (review.NamespaceReviewStatus == NamespaceReviewStatus.Approved || review.IsApproved))
+                        {
+                            // Handle reviews that are already approved - ensure they have the group information
+                            review.NamespaceReviewStatus = NamespaceReviewStatus.Approved;
```
why on this approved we are adding all the details (group id, user id, requested on)

but on line 452 that we are doing basically the same thing, the values are not updated?

helen229 replied:
if this review has been requested before and approved, the group id would be different, and in previous conditions those won't be updated, it will go wrong when we query by group id

AlitzelMendez commented on src/dotnet/APIView/APIViewWeb/Managers/ReviewManager.cs:
```diff
@@ -702,6 +761,11 @@ private async Task CheckAndApproveNamespaceForTypeSpec(ReviewListItemModel appro
 
+                // Only approve TypeSpec namespace if there are no more pending SDK reviews
+                if (pendingSdkReviews.Any())
+                {
```
you can more this to line 749 after getting pendingSdkReviews, makes a little bit more sense
or you can join this with the if of typeSpecReview to have something like 

```csharp
if (typeSpecReview == null || pendingSdkReviews.Any()) {
    return;
}
```

helen229 commented on src/dotnet/APIView/APIViewWeb/Managers/ReviewManager.cs:
```diff
@@ -406,23 +441,37 @@ private async Task MarkAssociatedReviewsForNamespaceReview(List<ReviewListItemMo
                             }
+
+                            await _reviewsRepository.UpsertReviewAsync(review);
+                        }
```
If the review is approved, the namespace is automatically approved too, but namespace approval (first release approve) is not same as review approval. My understanding is that IsApprove is a superset of namespace approval

AlitzelMendez commented:
I think this PR is growing more than it should. Right now, this pull request is fixing four issues, and at this point, it’s hard to split. But in the future—and even if you can do something in this pull request—it would be great if we could handle one pull request per issue. When I review it now, I need to think about four different workflows, which makes it harder to review every iteration, smaller prs are better :)

helen229 commented:
> Most of the issues listed here have been fixed:
> 
> * [Prevent overriding existing approved TypeSpec namespaces #12211](https://github.com/Azure/azure-sdk-tools/issues/12211)
> * [Request namespace button should be disabled after use per revision #12204](https://github.com/Azure/azure-sdk-tools/issues/12204)
> * [The feature doesn't support the "approve-first, request-later" workflow #12215](https://github.com/Azure/azure-sdk-tools/issues/12215)
> * [Hide Request button if no related APIs #12644](https://github.com/Azure/azure-sdk-tools/issues/12644)
> 
> However, other things are still broken, and honestly, I can’t continue reviewing a pull request that touches so many areas without proper test coverage. I see there are tests in the code, but most of them are not meaningful. I understand that right now we’re focusing on other priorities, but these kinds of failures are concerning:
> 
> Approved changes not being reflected: https://spa.apiviewuxtest.com/review/93790a28c3464c9f94ee3fefe8b7cff3?activeApiRevisionId=ed1f6f9f918046d9af34f3150be75fce Linking a ton of namespaces: https://spa.apiviewuxtest.com/review/d1d3c6098b994511a997be68146e7f57?activeApiRevisionId=4f05b0bc51b449589b4bad711371fcbc
> 
> This is not production-ready and definitely not merge-ready.
> 
> I really don’t want to review this pull request again without proper tests. Relying on manual testing gives me the impression that the code wasn’t tested before reaching me, and I don’t feel comfortable approving it.
> 
> To be fair, there are fixes in this pull request, such as:
> 
> * The button being disabled after a request — fixed.
> * The code considering previously approved items — also fixed.
> * Support approve first request later - also fixed
> 
> My impression is that one of the bugs was introduced by hide request button if not related pull requests and overall, this is another large pull request that touches too many things and breaks others along the way.


This issue is only introduced by my latest changes of DB, it wasn't before my latest commit. I have tested it in my local for checking the DB updates and haven't got chance to test in Ux-test for all flows before you did. Sorry about that.
Approved changes not being reflected: https://spa.apiviewuxtest.com/review/93790a28c3464c9f94ee3fefe8b7cff3?activeApiRevisionId=ed1f6f9f918046d9af34f3150be75fce


This is an existing issue in ApiView for quite long time, not related to any of my changes. But yes, we should fix this in recent sprints.
Linking a ton of namespaces: https://spa.apiviewuxtest.com/review/d1d3c6098b994511a997be68146e7f57?activeApiRevisionId=4f05b0bc51b449589b4bad711371fcbc


So far, we don't have end to end testing framework established in APIVIEW codebase, manual testing can't be avoided for mocking user behaviors and all kinds of scenarios. I can list test cases I have tested and results like we did in release planner release notes to make you feel better to continue reviewing and focus on code changes only.
